src/App.tsx

```diff
@@ -1,26 +1,216 @@
+    inputRef.current?.focus();
+  };
+
+  const handleToggleStatus = (todo: Todo) => {
+    setLoadingTodos(current => [...current, todo.id]);
+
+    updateTodo({ ...todo, completed: !todo.completed })
+      .then(updatedTodo =>
+        setTodos(currentTodos =>
+          currentTodos.map(currentTodo =>
+            currentTodo.id === updatedTodo.id ? updatedTodo : currentTodo,
+          ),
+        ),
+      )
+      .catch(() => setError(Error.UnableUppdateTodo))
+      .finally(() =>
+        setLoadingTodos(current => current.filter(id => id !== todo.id)),
+      );
+  };
+
+  const handleToggleAll = () => {
+    if (activeTodos.length > 0) {
+      activeTodos.forEach(handleToggleStatus);
+    } else {
+      completedTodos.forEach(handleToggleStatus);
+    }
+  };
+
+  const handleRename = (todo: Todo) => {
+    setLoadingTodos(current => [...current, todo.id]);
+
+    updateTodo(todo)
+      .then(updatedTodo =>
+        setTodos(currentTodos =>
+          currentTodos.map(currentTodo =>
+            currentTodo.id === updatedTodo.id ? updatedTodo : currentTodo,
+          ),
+        ),
+      )
+      .catch(() => setError(Error.UnableUppdateTodo))
+      .finally(() =>
+        setLoadingTodos(current => current.filter(id => id !== todo.id)),
+      );
+  };
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
   return (
-    <section className="section container">
-      <p className="title is-4">
-        Copy all you need from the prev task:
-        <br />
-        <a href="https://github.com/mate-academy/react_todo-app-add-and-delete#react-todo-app-add-and-delete">
-          React Todo App - Add and Delete
-        </a>
-      </p>
-
-      <p className="subtitle">Styles are already copied</p>
-    </section>
+    <div className="todoapp">
+      <h1 className="todoapp__title">todos</h1>
+
+      <div className="todoapp__content">
+        <Header
+          inputRef={inputRef}
+          todos={todos}
+          title={title}
+          onAdd={handleAddTodo}
+          newTitle={setTitle}
+          onToggleAll={handleToggleAll}
+        />
+        <TodoList
+          todos={filteredTodos}
+          onDelete={handleDeleteTodo}
+          tempTodo={tempTodo}
+          loadingTodos={loadingTodos}
+          onToggle={handleToggleStatus}
+          onRename={handleRename}
+        />
+
+        {todos.length !== 0 && (
+          <Footer
+            todos={todos}
+            selectedFilter={selectedFilter}
+            setSelectedFilter={setSelectedFilter}
+            onClearCompleted={handleDeleteAllCompleted}
+          />
+        )}
+      </div>
+
+      <div
+        data-cy="ErrorNotification"
+        className={cn(
+          'notification is-danger is-light has-text-weight-normal',
+          { hidden: !error },
+        )}
+      >
+        <button
+          data-cy="HideErrorButton"
+          type="button"
+          className="delete"
+          onClick={() => setError('')}
```

Comment:
[Not critical]
Clearing error function is used in multiple places, so it can be defined once as a separate function and called in all those places.